File: solutions/python/bit-manipulation/total-hamming-distance.py
# LeetCode 477 — Total Hamming Distance (Medium)
# Category: Bit Manipulation · Approach: Per-bit counting
# Time: O(32n) | Space: O(1)
# Source: https://leetcode.com/problems/total-hamming-distance/

def total_hamming_distance(nums: list[int]) -> int:
    n = len(nums)
    total = 0
    for bit in range(32):
        ones = sum((x >> bit) & 1 for x in nums)
        total += ones * (n - ones)
    return total

File: solutions/typescript/bit-manipulation/total-hamming-distance.ts
// LeetCode 477 — Total Hamming Distance (Medium)
// Category: Bit Manipulation · Approach: Per-bit counting
// Time: O(32n) | Space: O(1)
// Source: https://leetcode.com/problems/total-hamming-distance/

function totalHammingDistance(nums: number[]): number {
  const n = nums.length;
  let total = 0;
  for (let bit = 0; bit < 32; bit++) {
    let ones = 0;
    for (const x of nums) ones += (x >> bit) & 1;
    total += ones * (n - ones);
  }
  return total;
}
